crates/ide/src/goto_definition.rs

```diff
@@ -190,6 +202,200 @@ fn try_filter_trait_item_definition(
+        file_id: HirFileId,
+        ancestors: impl Iterator<Item = SyntaxNode>,
+        cursor_token_kind: SyntaxKind,
+    ) -> Option<UpmappingResult<NavigationTarget>> {
+        let db = sema.db;
+
+        for anc in ancestors {
+            match_ast! {
+                match anc {
+                    ast::Fn(fn_) => {
+                        let hir_fn: hir::Function = sema.to_def(&fn_)?;
+                        let nav = hir_fn.try_to_nav(db)?;
+
+                        // For async token, we navigate to itself, which triggers
+                        // VSCode to find the references
+                        let focus_token = if matches!(cursor_token_kind, T![async]) {
+                            fn_.async_token()?
+                        } else {
+                            fn_.fn_token()?
+                        };
+                        let focus_range = InFile::new(file_id, focus_token.text_range())
+                            .original_node_file_range_opt(db)
+                            .map(|(frange, _)| frange.range);
+
+                        return Some(nav.map(|it| {
+                            if focus_range.is_some_and(|range| it.full_range.contains_range(range)) {
+                                NavigationTarget { focus_range, ..it }
+                            } else {
+                                it
+                            }
+                        }));
+                    },
+                    ast::ClosureExpr(c) => {
+                        let pipe_tok = c.param_list().and_then(|it| it.pipe_token())?.into();
+                        let nav = NavigationTarget::from_expr(db, InFile::new(file_id, c.into()), pipe_tok);
+                        return Some(nav);
+                    },
+                    ast::BlockExpr(blk) => match blk.modifier() {
+                        Some(ast::BlockModifier::Async(_)) => {
+                            let async_tok = blk.async_token()?.into();
+                            let nav = NavigationTarget::from_expr(db, InFile::new(file_id, blk.into()), async_tok);
+                            return Some(nav);
+                        },
+                        Some(ast::BlockModifier::Try(_)) if cursor_token_kind != T![return] => {
+                            let try_tok = blk.try_token()?.into();
+                            let nav = NavigationTarget::from_expr(db, InFile::new(file_id, blk.into()), try_tok);
+                            return Some(nav);
+                        },
+                        _ => {}
+                    },
+                    _ => {}
+                }
+            }
+        }
+        None
+    }
+
+    let token_kind = token.kind();
+    sema.descend_into_macros(DescendPreference::None, token.clone())
+        .into_iter()
+        .filter_map(|descended| {
+            let file_id = sema.hir_file_for(&descended.parent()?);
+
+            // Try to find the function in the macro file
```

Comment:
I think you want to go via `Semantics::token_ancestors_with_macros` instead,

Reply:
Actually they are different. Here's an example:

```rust
macro_rules! M {
    ($blk:expr) => {
        fn f1() {
            $blk
        }

        $blk
    };
}

fn main() {
    M!({ return; /* goto-def on this return*/ });
}
```

When we perform on the return in the macro call-site, it should navigate to both `f1` and `main`. So we should take all the occurrence of `return` (both real files and macro files) into consideration.

Reply:
Yes but that is covered by iterating through all the `descend_into_macros` results. The `return` token should yield two tokens in `descend_into_macros`, and on each we'd call `token_ancestors_with_macros` which ought to yield us the 2 functions

Reply:
Oh, I get it. I misunderstood what you meant earlier.